### PR TITLE
add channel handler events

### PR DESCRIPTION
**Pull request by mekolat**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [X] I have followed [proper Hercules code styling][code].
- [X] I have read and understood the [contribution guidelines][cont] before making this PR.
- [X] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
* Allow chat channels to have event handlers
* Add `addchannelhandler("<#channel>", "<NPC::OnEvent>")`
* Add `removechannelhandler("<#channel>", "<NPC::OnEvent>")`
* Update the documentation accordingly

<br>

### Example
```c
OnChannelMessage:
    if (compare(@channelmes$, "hello")) {
        channelmes("#chan", "NPC : hi there!");
    }
    end;

OnInit:
    addchannelhandler("#chan", "NPC::OnChannelMessage");
```

<br><br>

**Affected Branches:** `master`
**Type:** [enhancement](https://github.com/HerculesWS/Hercules/labels/type%3Aenhancement)
**Compopents:** [core](https://github.com/HerculesWS/Hercules/labels/component%3Acore), [script engine](https://github.com/HerculesWS/Hercules/labels/component%3Acore%3Ascriptengine), [documentation](https://github.com/HerculesWS/Hercules/labels/component%3Adocumentation)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1719)

File: src/map/channel.c [hunk 1]
```diff
@@ -278,12 +279,21 @@ void channel_send(struct channel_data *chan, struct map_session_data *sd, const
...
 		safesnprintf(message, 150, "[ #%s ] %s : %s", chan->name, sd->status.name, msg);
 		clif->channel_msg(chan,sd,message);
 		if (chan->type == HCS_TYPE_IRC)
 			ircbot->relay(sd->status.name,msg);
 		if (chan->msg_delay != 0)
 			sd->hchsysch_tick = timer->gettick();
+
+		for (i = 0; i < MAX_EVENTQUEUE; i++) {
+			if (chan->handlers[i][0] != '\0') {
+				script->set_var(sd, "@channelmes$", aStrdup(msg));
```
Comment (MishimaHaruna) [inline src/map/channel.c]: This seems a memory leak: `script->set_var()` doesn't take ownership of the string you pass (it takes a const pointer), but creates its own copy internally. The call sequence is:

1. `script_set_var()` calls `script->set_reg()`, passing the const pointer as is (`value`) and passes no reg_db (`ref == NULL`)
2. `script_set_reg()` calls `pc->setregstr()` if the variable if of type string and `ref` is NULL. Once again, it passes the const pointer as is (casts from `const void *` to `const char *`.
3. `pc_setregstr()` checks that the string isn't empty (zero length). If so, it duplicates it through `aStrdup()`.

You can simply pass `msg`, without duplicating it.

**Comment by mekolat** [comment]
@MishimaHaruna better now?

**Comment by MishimaHaruna** [comment]
Perfect, thank you!

**Pull request merged**